pkg/controller/apiextensions/composite/api.go

```diff
@@ -301,3 +307,49 @@ func (c *APINamingConfigurator) Configure(ctx context.Context, cp resource.Compo
 	meta.AddLabels(cp, map[string]string{composed.LabelKeyNamePrefixForComposed: cp.GetName()})
 	return errors.Wrap(c.client.Update(ctx, cp), errUpdateComposite)
 }
+
+// NewAPIPrioritizedDeleter returns a new *APIPrioritizedDeleter.
+func NewAPIPrioritizedDeleter(client client.Client) *APIPrioritizedDeleter {
+	return &APIPrioritizedDeleter{client: client}
+}
+
+// APIPrioritizedDeleter deletes the composed objects in layers that are specified
+// as deletion priority in the *Composition object. Deletion priority of a composed
+// object determines its layer and deletion waits for higher layers to complete
+// before proceeding to the lower one.
+type APIPrioritizedDeleter struct {
+	client client.Client
+}
+
+// Delete makes sure the current highest priority is deleted and the remaining
+// ojbects are returned.
+func (d *APIPrioritizedDeleter) Delete(ctx context.Context, cr resource.Composite, comp *v1alpha1.Composition) ([]resource.Composed, error) {
+	var del []resource.Composed
+	hp := int64(math.MinInt64)
+	for i, ref := range cr.GetResourceReferences() {
+		u := runtimecomposed.New(runtimecomposed.FromReference(ref))
+		nn := types.NamespacedName{Name: ref.Name, Namespace: ref.Namespace}
+		err := d.client.Get(ctx, nn, u)
+		if resource.IgnoreNotFound(err) != nil {
+			return nil, errors.Wrap(err, errGetComposed)
+		}
+		if kerrors.IsNotFound(err) {
+			continue
+		}
```

Comment:
Not sure if worth it, especially initially, but it would be nice to not have to `Get` resources on subsequent reconciles we could know to have already been deleted.

Reply:
Also, we `Get` every resource here even if the known deletion priority of the resource will not result in it being deleted on this reconcile. We could sort resources beforehand by deletion priority, then iterate through and as soon as we `Get` a resource that exists then we can exit the loop if the next resource's deletion priority is less than that of the current. An even simpler solution to start with would be to move the check below above the `Get` call so that if `hp` is already greater then this resource's `p` (i.e. we have gotten a resource that exists and has a higher deletion priority), then we can skip getting this resource for this reconcile.

Reply:
> but it would be nice to not have to Get resources on subsequent reconciles we could know to have already been deleted.

Between reconciles, we have only the custom resource itself as the state store that we can store at which level we were in the former reconcile, but I'm not sure if adding a field on composite resource is worth it. Keep in mind that most of these `Get` calls are fulfilled by the cache. Also, we can't remove an entry from `spec.resourceRefs` since the order of those resources correspond to the order of the templates in the `Composition` due to the fact that we don't name the templates in `Composition`, so index is used as identifier.

> We could sort resources beforehand by deletion priority, then iterate through and as soon as we Get a resource that exists then we can exit the loop if the next resource's deletion priority is less than that of the current.

Yeah I've tried sorting and it looks like this in practice:
```go
layers := map[int64][]corev1.ObjectReference{}
for i, ref := range cr.GetResourceReferences() {
	layers[comp.Spec.Resources[i].DeletionPriority] = append(layers[comp.Spec.Resources[i].DeletionPriority], ref)
}
type layer struct {
	refs     []corev1.ObjectReference
	priority int64
}
var list []layer
for key, val := range layers {
	list = append(list, layer{refs: val, priority: key})
}
sort.Slice(list, func(i, j int) bool {
	return list[i].priority > list[j].priority
})
cp := int64(math.MaxInt64)
var del []corev1.ObjectReference
for _, layer := range list {
	for _, ref := range layer.refs {
		u := runtimecomposed.New(runtimecomposed.FromReference(ref))
		err := d.client.Delete(ctx, u)
		if resource.IgnoreNotFound(err) != nil {
			return nil, errors.Wrap(err, errDeleteComposed)
		}
		if kerrors.IsNotFound(err) {
			continue
		}
		cp = layer.priority
	}
	if cp == layer.priority {
		del = layer.refs
		break
	}
}
return del, nil

```

So, it's less api-server calls but I think the logic is more complicated with many loops. Also, as soon as you iterate with `Get` calls, I don't think there is a stark difference between going through each object vs only until the current priority level given that the length of resources array is small most of the time.

> An even simpler solution to start with would be to move the check below above the Get call so that if hp is already greater then this resource's p (i.e. we have gotten a resource that exists and has a higher deletion priority), then we can skip getting this resource for this reconcile.

I'm not sure I got that. How are we able to know that resource exists without making either `Get` or `Delete` call? The reason I do a `Get` is that a resource with higher priority could always be chosen even if it's been accounted for and deleted, so, we filter those via the `Get` call.

Reply:
BTW, feel free to post pseudo code if I didn't understand what you mean correctly.

Reply:
for the simple solution I just mean something like this prior to the `Get` call:
```go
p := comp.Spec.Resources[i].DeletionPriority
if p < hp { // if we have already found a resource that exists and has higher deletion priority we don't need to GET this one
  continue
}
```